### PR TITLE
Remove `delete_existing` option; Introduce archive

### PR DESCRIPTION
**Pull request by azrv**
Remove `delete_existing` run option, instead archive memory and workspace on every run;

`archive` is a step since some flows don't generate code and therefore should not archive whatever

Relates to https://github.com/AntonOsika/gpt-engineer/issues/388

### CODE REVIEW

File: gpt_engineer/db.py [hunk 1]
```diff
@@ -47,3 +48,4 @@ class DBs:
     preprompts: DB
     input: DB
     workspace: DB
+    archive: Optional[DB]
```
Comment (AntonOsika) [inline gpt_engineer/db.py:51]: Why optional?

```suggestion
    archive: DB
```

File: gpt_engineer/main.py [hunk 2]
```diff
@@ -35,33 +34,32 @@ def main(
...
         workspace=DB(workspace_path),
         preprompts=DB(Path(__file__).parent / "preprompts"),
+        archive=None if initial_run else DB(archive_path),
     )
 
     steps = STEPS[steps_config]
     for step in steps:
         messages = step(ai, dbs)
+        # TODO: Don't dump if step returns None
+        # human_review, archive don't need to be dumped
```
Comment (AntonOsika) [inline gpt_engineer/main.py:62]:
```suggestion
```

File: gpt_engineer/main.py [hunk 3]
```diff
@@ -35,33 +34,32 @@ def main(
...
         input=DB(input_path),
         workspace=DB(workspace_path),
         preprompts=DB(Path(__file__).parent / "preprompts"),
+        archive=None if initial_run else DB(archive_path),
     )
 
     steps = STEPS[steps_config]
     for step in steps:
         messages = step(ai, dbs)
+        # TODO: Don't dump if step returns None
```
Comment (AntonOsika) [inline gpt_engineer/main.py:61]: It's OK to write empty files...

```suggestion
```

File: gpt_engineer/steps.py [hunk 4]
```diff
@@ -258,6 +261,18 @@ def fix_code(ai: AI, dbs: DBs):
     return messages
 
 
+def archive(ai: AI, dbs: DBs):
+    if dbs.archive is None:
```
Comment (AntonOsika) [inline gpt_engineer/steps.py:265]:
```suggestion
```

File: gpt_engineer/steps.py [hunk 5]
```diff
@@ -258,6 +261,18 @@ def fix_code(ai: AI, dbs: DBs):
...
+        return
```
Comment (AntonOsika) [inline gpt_engineer/steps.py:266]:
```suggestion
```

File: gpt_engineer/steps.py [hunk 6]
```diff
@@ -280,15 +295,17 @@ class Config(str, Enum):
 # Different configs of what steps to run
 STEPS = {
     Config.DEFAULT: [
+        archive,
```
Comment (AntonOsika) [inline gpt_engineer/steps.py:298]: Let's not make archive a step in this sense, but just call it before we loop over steps!

**Comment by AntonOsika** [comment]
Great work!

Added some comments.

Mainly I think we shouldn't let archive be a "step" but we just call it.

Also, feel free to delete the "run_id" functionality.

**Review by AntonOsika — CHANGES_REQUESTED** [review]

**Comment by azrv** [comment]
> Great work!
> 
> Added some comments.
> 
> Mainly I think we shouldn't let archive be a "step" but we just call it.
> 
> Also, feel free to delete the "run_id" functionality.

@AntonOsika 

1) My main concern about archiving NOT being a step is there are "flows" that don't generate code (e.g. EXECUTE_ONLY, EVALUATE) and therefore should not archive things by default. How do we make sure to skip archiving in those cases?
While working on PR I realised archiving functionality complicates code a lot and I'm no sure it provides us any good.
What's inherently wrong with `delete_existing`?

2) Sure! Could you please elaborate on what `run_id` is
I can't seem to find it (not so familiar with the codebase yet) 
`git grep run_id` gives nothing

**Comment by AntonOsika** [comment]
Merging!

**Comment by AntonOsika** [comment]
Thanks for the submission @azrv !

Do you want to join discord and DM me there? Happy to chat about future improvements!

**Comment by AntonOsika** [comment]
Already cleaned up run_id (was callsed run_prefix)

**Pull request merged**